Uses PrintErr to communicate error messages

data-bomb opened the pull request:
Switches from raise exception to printerr to communicate error messages. There are situations where a signature cannot be found and the printerr displays the error message in red text while withholding the additional information triggered by raising an exception.

data-bomb committed:
Uses PrintErr to communicate error messages

-Switches from raise exception to printerr to communicate error messages

nosoop commented on makesig.py:
```diff
@@ -136,7 +135,7 @@ def process(start_at = MAKE_SIG_AT['fn']):
 	fn = fm.getFunctionContaining(currentAddress)
 	if not fn:
-		raise Exception("Not in a function")
+		printerr("Not in a function")
```
Raising an exception here (and subsequently breaking out of the script) makes sense, since this script doesn't generate signatures for arbitrary bytes.

data-bomb replied:
Fixed in https://github.com/nosoop/ghidra_scripts/pull/5/commits/b3f0a71cbebf5a0140f61b36ae79f40dba971936 The logic wasn't corrected here from another version. Much of this PR may come down to personal preferrence on what we want to see in these scenarios. An exception being thrown generates a lot of other information which didn't seem as helpful as the error message in correcting user behavior.

nosoop commented on makesig.py:
```diff
@@ -105,10 +105,9 @@ def process(start_at = MAKE_SIG_AT['fn']):
-			raise Exception("Instruction at %s is not adjacent"
+			printerr("Instruction at %s is not adjacent"
 					" to previous (expected %s)" % (expected_next, ins.getAddress()))
+			break
```
Is this supposed to result in not finding a unique signature? My immediate thought is that it'd be better to fill in discontiguous bytes with wildcards, but in practice I'm not sure.

data-bomb replied:
If it's possible to have a chance of getting a signature by wildcarding them until we get out of no-man's-land then that'd be great. May need to have a threshold of number of instructions before we abandon all hope?

nosoop replied:
Wildcarding `ins.GetAddress() - expected_next` bytes may suffice; just peeked and [the `Instruction.getNext()` docs indicate](https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Instruction.html#getNext()) that it's the one that follows in address order.

data-bomb replied:
That worked for me with some limited testing. Proposed change in https://github.com/data-bomb/makesig/commit/ba7151a0fb1882a0d1af51b212f7e141144234fd. It may not get exercised a lot but it should be helpful. I also tried removing the safeguards of remaining in the same function to confirm it would work in this scenario as well.

nosoop reviewed (commented):
The original intent is that raising an exception halts script execution as well; it doesn't seem like `printerr` does.

data-bomb committed:
Fixes logic bug in main

Fixes logical issue with a printerr call in main which would still start process() if we're not in a function

data-bomb committed:
Adds wildcards between valid instructions

Removes printerr in favor of adding wildcards to try and reach the next valid instruction

nosoop commented:
This should do.  Thanks!
